cli/sql: use BSD's libedit for input.

knz opened the pull request:
*Very many special thanks to @benesch and @bdarnell without whom this never could have been possible.*


Fixes #18192.
Fixes #16324.
Fixes #15460.
Fixes #14938.
Fixes #18909.

### cli/sql: use BSD's libedit for input.

This patch steps over to use http://github.com/knz/go-libedit,
a Go wrapper around BSD's
[editline](https://www.freebsd.org/cgi/man.cgi?query=editline&sektion=3)
library (`libedit`).

This supports:

- custom key bindings via `~/.editrc`
  See https://www.freebsd.org/cgi/man.cgi?query=editrc
  for details.
- proper entry across multiple lines.
- all terminal types supported by the system's terminfo library.

The name of the history file is changed from `~/.cockroachdb_history`
to `~/.cockroachsql_history` because the format is different. This
both avoids an error upon loading the previous format in the new
shell, and garbage history when trying to run an old shell over the
file using the new format.

### cli/sql: remove the option normalize_history

Prior to this patch, the CLI shell implemented logic to populate the
shell history using pretty-printed renderings of the AST of
parsed statements so far. The reason why this was done was a
combination of two factors:

- the readline input library was not able to handle multi-line
  entries.
- for UX it is desirable to have one "item" of SQL entry, that is,
  a full statement, as a single item of history so as to ease
  recollection of the history entry.

So to satisfy the 2nd point while dealing with the first, the shell
would concatenate the lines of a single statement, parse them,
then pretty-print the AST, because the pretty-print code guarantees
that the result contains a single line (significant newline characters
e.g. in string literals are escaped).

The second requirement still holds, but now the new readline library
*is* able to handle multi-line history entry.

So this patch simplifies the code to add multi-line statements
as one history entry, and removes the complexity associated with
normalization entirely.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18531)
<!-- Reviewable:end -->

knz commented:
@benesch I'll need your help to 

1) explain the builder how to generate linux binaries against a non-versioned libtinfo
2) vendor the dependency

knz commented:
I've done the vendoring but somehow teamcity does not agree with me. Not sure why.

benesch commented:
This seems to be the error:

```
[22:13:48][Step 2/3] In file included from vendor/github.com/knz/go-libedit/src/libedit-terminal.c:1:0,
[22:13:48][Step 2/3]                  from wrap-terminal.c:1:
[22:13:48][Step 2/3] vendor/github.com/knz/go-libedit/src/c-libedit/terminal.c:58:21: fatal error: termcap.h: No such file or directory
[22:13:48][Step 2/3]  #include <termcap.h>
[22:13:48][Step 2/3]                      ^
[22:13:48][Step 2/3] compilation terminated.
```

knz commented:
Okay I fixed the last issues with the Go code.
Blockers:
- #18687 - `dep` is making me unhappy for seemingly unrelated issues
- the builder image needs to be populated with the terminfo header files.

benesch commented:
Ah, so that indicates we've introduced a build-time dependency on `libterminfo-dev`. That'll need to be documented in the build from source instructions.

knz commented:
- the dep issue is solved
- the builder image has all the good files, no need to add more
- the remaining issue is that the acceptance tests do not use the builder image to build cockroachdb, and instead use our "acceptance test image". That image is missing some files (looking into it). I am somewhat confused as to why the acceptance tests do not use the builder to build though.

knz commented:
Okay so the builder has the terminfo lib and headers in /usr/include and /usr/lib, but the `make build` target uses the sysroot in /x-tools exclusively.

We need two things to happen, for the linux targets:
- symlink /usr/include/termcap.h into the corresponding /x-tools sysroots
- import the non-versioned libtinfo.so (e.g. from CentOS) and make the linker use that.

@benesch I'd like to sit down a few minutes with you to get this sorted.

knz commented:
Ok progress stalled once again due to `dep` breakage. Awaiting for update by @dt to rebase/regenerate.

knz commented:
Will rely on  #19007.

knz commented:
Builder updated; confirmed to generate working binaries with TYPE= `release-linux-gnu`, `release-linux-musl`, `portable`, `release-windows` and `release-darwin`.

Now waiting on `honnef.co` to resuscitate to update the vendor tree again.

knz committed:
cli/sql: use BSD's libedit for input.

This patch steps over to use github.com/knz/go-libedit,
a Go wrapper around BSD's
[editline](https://www.freebsd.org/cgi/man.cgi?query=editline&sektion=3)
library (`libedit`).

This supports:

- custom key bindings via `~/.editrc`
  See https://www.freebsd.org/cgi/man.cgi?query=editrc
  for details.
- proper entry across multiple lines.
- all terminal types supported by the system's terminfo library.

The name of the history file is changed from `~/.cockroachdb_history`
to `~/.cockroachsql_history` because the format is different. This
both avoids an error upon loading the previous format in the new
shell, and garbage history when trying to run an old shell over the
file using the new format.

bdarnell commented:
So the net result is that on mac/bsd, this links dynamically against libedit, on linux-glibc it embeds libedit and links dynamically against libtinfo, and on linux-musl it links everything statically, right? Does the linux-musl build need a terminfo database on disk or does it  have something embedded (for at least the most popular terminals)?

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[CONTRIBUTING.md, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18531#-KvdKZ0ZMeDNxf1ZMR4l:-KvdKZ0ZMeDNxf1ZMR4m:bde5z6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a2b1fc8e9adfe22ee492cf7420ab5985d80cbf9/CONTRIBUTING.md#L41)):*
> ```Markdown
>        buffers.
>      - On GNU/Linux specifically, the terminfo/ncurses development
>        headers (note: not needed on BSD, Windows and macOS). This is
> ```

On bsd/mac, they're still needed, but they're present by default, right?

---

*[pkg/cli/interactive_tests/test_history.tcl, line 11 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18531#-KvdL7IsQDPMCBp2z2vZ:-KvdL7IsQDPMCBp2z2v_:b-puqa01) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a2b1fc8e9adfe22ee492cf7420ab5985d80cbf9/pkg/cli/interactive_tests/test_history.tcl#L11)):*
> ```Tcl
> 
> # Ensure the connection is up and working.
> send "SELECT 1;\r"
> ```

Why are these now capitalized? Users will usually type in lowercase. If the history is now case-sensitive then the fact that we normalize to uppercase will be really annoying. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18531#-:-KvdMe7N2PvspW0tsWwJ:b-f6acyn)*
<!-- Sent from Reviewable.io -->

knz commented:
your analysis is correct. Additionally on windows and other unknown systems it falls back to a dumb stdin readline.

The static build needs a terminfo database on disk, however I checked and Alpine does indeed have one as part of the base system. Also the user can populate entries in their home directory in ~/.terminfo. There's an option to embed some entries but I haven't felt the need to do so yet.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[CONTRIBUTING.md, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18531#-KvdKZ0ZMeDNxf1ZMR4l:-KvdPmyJmsKAYHjb1PJ2:b-r960f8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a2b1fc8e9adfe22ee492cf7420ab5985d80cbf9/CONTRIBUTING.md#L41)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

On bsd/mac, they're still needed, but they're present by default, right?
</blockquote></details>

This comment is stale. I changed the code to not need any dev headers at all. Will update.

---

*[pkg/cli/interactive_tests/test_history.tcl, line 11 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18531#-KvdL7IsQDPMCBp2z2vZ:-KvdPqYMtylkXwziL_3R:b-tuq6z6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a2b1fc8e9adfe22ee492cf7420ab5985d80cbf9/pkg/cli/interactive_tests/test_history.tcl#L11)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why are these now capitalized? Users will usually type in lowercase. If the history is now case-sensitive then the fact that we normalize to uppercase will be really annoying. 
</blockquote></details>

Yes I just found this out. Will send out a separate PR in front of this one to disable normalization in the history. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18531#-:-KvdPPBuluPlxzkVvsBN:blr3v8l)*
<!-- Sent from Reviewable.io -->

knz committed:
cli/sql: remove the option normalize_history

Prior to this patch, the CLI shell implemented logic to populate the
shell history using pretty-printed renderings of the AST of
parsed statements so far. The reason why this was done was a
combination of two factors:

- the readline input library was not able to handle multi-line
  entries.
- for UX it is desirable to have one "item" of SQL entry, that is,
  a full statement, as a single item of history so as to ease
  recollection of the history entry.

So to satisfy the 2nd point while dealing with the first, the shell
would concatenate the lines of a single statement, parse them,
then pretty-print the AST, because the pretty-print code guarantees
that the result contains a single line (significant newline characters
e.g. in string literals are escaped).

The second requirement still holds, but now the new readline library
*is* able to handle multi-line history entry.

So this patch simplifies the code to add multi-line statements
as one history entry, and removes the complexity associated with
normalization entirely.

knz commented:
ready, PTAL

benesch commented:
This works impressively well! One small nit. On a machine that has a go-readline history file, this warning occurs:

```
cannot load the command-line history (file corrupted?): read_history: invalid argument
the history file will be cleared upon first entry
```

Perhaps we should use a new name for the history file? It's especially bad in the other direction, since go-readline mis-decodes libedit's history:

```
insert\040into\040db.t\040values\040(\012\040\040'a'\012);\012
```

On second thought, it might be pretty easy to migrate the history file. (Third thought: holy cow the escaping routines in libedit are _complicated_.)

---

Reviewed 6 of 8 files at r1, 2 of 2 files at r2.
Review status: 6 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18531#-:-KvhSUHi59RUAUQvqgTw:bc1hwof)*
<!-- Sent from Reviewable.io -->

knz commented:
- Restored CONTRIBUTING.md (ncurses headers are not necessary).
- Changed the history file name to become `~/.cockroachsql_history` (instead of `~/.cockroachdb_history` as previously, to avoid running into the error message)

benesch commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 4 of 4 files at r3.
Review status: 6 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18531#-:-KvhclPjhCvWdY0uLGpl:b-uwkowd)*
<!-- Sent from Reviewable.io -->

knz commented:
Yay! it's green! Shall we press the big green button?

knz commented:
Updated to take the fix to https://github.com/knz/go-libedit/issues/1.

jseldess commented:
@knz, will this be cherrypicked into 1.1? If so, I'll remove the associated known limitation from the 1.1 docs.

bdarnell commented:
No, this is far too big a change to cherry-pick.

jseldess commented:
OK, thanks, @bdarnell. I'll wait until it shows up in a patch release (or 1.2) before removing the limitations then.

knz commented:
Agreed this will need a few months of experimental testing by the team before we are happy about its stability.
